packages/runtime/src/state/modules/element-trees.ts

```diff
@@ -0,0 +1,358 @@
+        `Mismatching document keys ${documentKey} !== ${newDocument.key}`,
+      )
+
+      const elementTree = state.get(documentKey)
+      if (elementTree == null) return state
+
+      const updatedElementTree = applyChanges(
+        elementTree,
+        descriptors,
+        { old: oldDocument.rootElement, new: newDocument.rootElement },
+        operation,
+      )
+
+      return new Map(state).set(documentKey, updatedElementTree)
+    }
+
+    default:
+      return state
+  }
+}
+
+export function* traverseElementTree(
+  element: Element,
+  descriptors: DescriptorsByComponentType,
+): Generator<Element> {
+  yield element
+  if (isElementReference(element)) return
+
+  const elementDescriptors = descriptors.get(element.type)
+  if (elementDescriptors == null) return
+
+  for (const [propKey, descriptor] of Object.entries(elementDescriptors)) {
+    const children = Introspection.getElementChildren(descriptor, element.props[propKey])
+    for (const child of children) {
+      yield* traverseElementTree(child, descriptors)
+    }
+  }
+}
+
+function getElementIdProp(
+  element: ElementData,
+  descriptors: DescriptorsByComponentType,
+): string | null {
+  const elementDescriptors = descriptors.get(element.type)
+  if (elementDescriptors == null) return null
+
+  for (const [propName, descriptor] of Object.entries(elementDescriptors)) {
+    const elementId = Introspection.getElementId(descriptor, element.props[propName])
+    if (elementId != null) return elementId
+  }
+
+  return null
+}
+
+export function buildElementTree(
+  rootElement: Element,
+  descriptors: DescriptorsByComponentType,
+): ElementTree {
+  const elements = new Map<string, Element>()
+  const elementIds = new Map<string, string>()
+
+  for (const element of traverseElementTree(rootElement, descriptors)) {
+    elements.set(element.key, element)
+    if (!isElementReference(element)) {
+      const elementId = getElementIdProp(element, descriptors)
+      if (elementId != null) elementIds.set(element.key, elementId)
+    }
+  }
+
+  return {
+    elements,
+    elementIds,
+  }
+}
+
+type OperationPath = Operation[number]['p']
+
+// performance-sensitive function, intentionally not using `elementData` schema here
+function isElement(item: unknown): item is Element {
+  return (
+    typeof item === 'object' &&
+    item != null &&
+    'key' in item &&
+    'type' in item &&
+    typeof item.key === 'string' &&
+    typeof item.type === 'string'
+  )
+}
+
+type ElementOperationPath = { elementPath: OperationPath; propName: string }
+
+function getElementOperationPath(path: OperationPath): ElementOperationPath {
+  const i = path.findLastIndex(
+    (fragment, i) =>
+      typeof fragment === 'number' && (i === path.length - 1 || path[i + 1] === 'props'),
+  )
+
+  if (path.length === 0 || (i === -1 && path[0] !== 'props') || path.length - i < 3) {
+    console.error('Operation path does not point to an element property', { path })
```

Comment:
I'm curious what you think about handling these errors as a first-class thing at some point in the future?